### PR TITLE
Mastery Mismatch Fix

### LINKED ISSUE #2087: exercise is not completed even after solving enough questions correctly in a row
## Summary

Level is not completed even after solving 4-5 questions correctly in a row.

## Detailed Description

- Level is not completed even after solving 4-5 questions correctly in a row.
- After a certain question, once you click on check answer tick comes ,Post that if you click  on next question same question and tick appear in a loop. The exercise shows still in progress.
- The problem is solved only when you refresh the page. It shows a green tick and next item after this.
- Not everyone but most of the learners facing this problem.

## System information

- Kolibri Version: Kolibri 0.4.5
- Operating System : Ubuntu 14.04
- Browser : Chrome

## How to reproduce

1. start giving exercise with learner login

## Video
[VID-20170809-WA0007.mp4.zip](https://github.com/learningequality/kolibri/files/1269360/VID-20170809-WA0007.mp4.zip)
[Error_Grade4.mp4.zip](https://github.com/learningequality/kolibri/files/1270056/Error_Grade4.mp4.zip)
[Error_Grade5.mp4.zip](https://github.com/learningequality/kolibri/files/1270055/Error_Grade5.mp4.zip)

### LINKED ISSUE #2089: detailed progress is sometimes missing under the 'Recent' coach report
## Summary

In the Recent Coach Report feature, the progress of some students is shown in percentage but a detailed question wise report does not appear.

## Detailed Description

- In the recent coach report, we saw students named Akash Shah have progress reported as 100 percent complete.
- After clicking on the name and looking at the report of the student, a detailed report was not being displayed. (Look at the Video).

## System information

Kolibri Version: Kolibri 0.4.5
Operating System : Ubuntu 14.04
Browser : Chrome

## How to reproduce

1. Create class scanario.
2. Start giving exercise through learner login.

## Video
[VID-20170901-WA0001.mp4.zip](https://github.com/learningequality/kolibri/files/1269460/VID-20170901-WA0001.mp4.zip)

### LINKED ISSUE #2252: accumulated points reduces suddenly on versions 0.4.5 and 0.5.0.
## Summary
The points accumulated by learners reduces suddenly on versions 0.4.5 and 0.5.0.

**This happens all of a sudden a learner with 6500 suddenly reduces to 4500 and 22,500 suddenly reduces to 12,500 etc.**

## System information
 - Version: 0.4.5 and 0.5.0
 - Operating system: Ubuntu 14.04LTS
 - Browser: Chrome

## How to reproduce
This is erratic and the best shot is to keep solving the exercise and observe the change in points.

## Screenshots
Sorry could not capture this since it happens at random intervals.

## Real-life consequences (anything community should be aware of, for instance how it affects your deployment)
1. Heartbroken learner who was religiously competing in the classroom to come first.
2. Disappointed coach.

### LINKED ISSUE #2253: 'nice work' pop-up often doesn't appear
## Summary
The nice work keep it up pop up doesn't come in most of the learners account on version 0.4.5 and 0.5.0

## System information
 - Version: 0.4.5 and 0.5.0
 - Operating system: Ubuntu 14.04LTS
 - Browser: Chrome

## How to reproduce
Start solving the exercise and videos the pop up does not come up.

## Screenshots
This pop up does not appear most of the time:
popup

## Real-life consequences (anything community should be aware of, for instance how it affects your deployment)

Learners are unaware that they have mastered and keep solving it continuously.

### PR DESCRIPTION
**Pull request by DXCanas**
## Summary

Bad times were had in the making of this PR. Fixes #2253, resolves #2089, fixes #2087, fixes #2252

### CODE REVIEW

**Comment by DXCanas** [comment]
@rtibbles In my tests, clearing out the log objects at page load for the `show*Content` pages just broke everything until this change was made. Looks like just this change is sufficient, but we can do those wipes (like we talked about) for good measure if you like.

File: kolibri/plugins/learn/assets/src/views/assessment-wrapper/index.vue [hunk 1]
```diff
@@ -323,10 +323,6 @@ oriented data synchronization.
         this.$emit('stopTracking', ...args);
       },
     },
-    beforeDestroy() {
-      // Make sure any unsaved data is captured before tear down.
-      this.saveAttemptLogMasterLog();
```
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/views/assessment-wrapper/index.vue]: Yeah, I guess the problem here is that we just want to save the data, whereas this action saves the data, and then sets it.

I agree with @indirectlylit's comments on #2355, that we should avoid use of beforeDestroy, so I am wondering if we could try to achieve this same 'data persistence cleanup' in the actions instead.

**Comment by rtibbles** [comment]
A clean state would be nice - but my main concern with the current implementation is that we could lose data (we have limited the amount of logging that occurs for performance reasons, so things could get lost by not doing a final save).

**Comment by DXCanas** [comment]
orderofevents

Just some confirmation of the race condition

**Comment by DXCanas** [comment]
@rtibbles updated!

**Pull request merged**